handler/src/main/java/io/netty/handler/ssl/JettyAlpnSslEngine.java

```diff
@@ -27,46 +28,61 @@
 import javax.net.ssl.SSLHandshakeException;
 
 import io.netty.util.internal.PlatformDependent;
+import org.conscrypt.OpenSSLEngineImpl;
 import org.eclipse.jetty.alpn.ALPN;
-import org.eclipse.jetty.alpn.ALPN.ClientProvider;
-import org.eclipse.jetty.alpn.ALPN.ServerProvider;
 
-final class JdkAlpnSslEngine extends JdkSslEngine {
-    private static boolean available;
+abstract class JettyAlpnSslEngine extends JdkSslEngine {
+    private static final boolean available = initAvailable();
 
     static boolean isAvailable() {
-        updateAvailability();
         return available;
     }
 
-    private static void updateAvailability() {
+    private static boolean initAvailable() {
         // TODO: Add support for ALPN when using Java9 and still be able to configure it the Netty way.
-        if (available || PlatformDependent.javaVersion() > 8) {
-            return;
+        if (PlatformDependent.javaVersion() <= 8) {
+            try {
+                // Always use bootstrap class loader.
+                Class.forName("sun.security.ssl.ALPNExtension", true, null);
+                return true;
+            } catch (Throwable ignore) {
+                // alpn-boot was not loaded.
+            }
         }
+        return false;
+    }
 
-        try {
-            // Always use bootstrap class loader.
-            Class.forName("sun.security.ssl.ALPNExtension", true, null);
-            available = true;
-        } catch (Exception ignore) {
-            // alpn-boot was not loaded.
-        }
+    static JettyAlpnSslEngine newClientEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ClientEngine((OpenSSLEngineImpl) engine, applicationNegotiator);
     }
 
-    JdkAlpnSslEngine(SSLEngine engine, final JdkApplicationProtocolNegotiator applicationNegotiator, boolean server) {
+    static JettyAlpnSslEngine newServerEngine(SSLEngine engine,
+            JdkApplicationProtocolNegotiator applicationNegotiator) {
+        return new ServerEngine((OpenSSLEngineImpl) engine, applicationNegotiator);
```

Comment:
casting to `OpenSSLEngineImpl` looks wrong here ... hopefully unit tests catch this?